### DISCUSSION TITLE
Curl hangs on some sites when using `--head`

### DISCUSSION THREAD

**Discussion by GlassGruber** [curl/curl#14670] [Q&A] [answered]
Hei there! I think, though I'm not sure, that this issue is caused by a wrong configuration of target servers that's why I'm not opening a ticket. 
When using the `--head` flag for the request the connection hangs indefinitely, I've left it more than 2min then cancelled manually.
But when making the request without `--head` option the connection completes normally with no hiccups.

This is the simple command used `curl -L --head <url>`, below a few urls with this behavior. 

```
http://scms.it
http://www.hungsingmartialarts.com/
http://www.aactg.it/
http://www.kungfurovato.com/
```

```txt
$ curl --version
curl 8.9.1 (x86_64-w64-mingw32) libcurl/8.9.1 LibreSSL/3.9.2 zlib/1.3.1 brotli/1.1.0 zstd/1.5.6 WinIDN libpsl/0.21.5 libssh2/1.11.0 nghttp2/1.62.1 ngtcp2/1.6.0 nghttp3/1.4.0
Release-Date: 2024-07-31
Protocols: dict file ftp ftps gopher gophers http https imap imaps ipfs ipns ldap ldaps mqtt pop3 pop3s rtsp scp sftp smb smbs smtp smtps telnet tftp ws wss
Features: alt-svc AsynchDNS brotli HSTS HTTP2 HTTP3 HTTPS-proxy IDN IPv6 Kerberos Largefile libz NTLM PSL SPNEGO SSL SSPI threadsafe UnixSockets zstd
```

**Comment by icing** [comment]
Looked at http://scms.it/, which 301 redirects to https://scms.it/. It then answers HTTP/2 with a 503 and headers, but never closes the response body. That is why curl "hangs", it waits for the response stream to be closed by the server.

With `--http1.1` it does not hang, since in 1.1 the end of the response requires no signal and is deduced by carrying neither `Content-Length` nor a `Transfer-Encoding: chunked`.

**Reply by icing** [reply]
I believe either the nginx configuration or its HTTP/2 implementation is at fault here. Without the `--head` the server sends the response body and closes the stream correctly.

**Reply by GlassGruber** [reply]
Yea that's what I thought as well. 
Naive question: why not just disconnect after the end of headers? After the headers empty line is transmitted isn't it "safe" to just drop the connection since what we want has been received?

**Reply by icing** [reply]
I'll have a look to see if we can do something about this.

**Comment by icing** [comment, marked as answer]
I made #14685 which fixes this. Please verify if you can.

**Reply by GlassGruber** [reply]
Sure thing! But I'm not sure where I can find a build artifact, I see there are daily builds but is not yet current with this patch I suppose.

**Reply by icing** [reply]
Ah, ok. No worries then.

**Comment by bagder** [comment]
#14685 was merged

**Reply by GlassGruber** [reply]
I've tested the daily [windows build](https://github.com/curl/curl-for-win/actions/runs/10603983606) builds and it works!  🎉
